man/sysupdate.d.xml

```diff
@@ -0,0 +1,872 @@
+      suggests).</para>
+
+      <para><programlisting># /usr/lib/sysupdate.d/60-root.conf
+[Transfer]
+ProtectVersion=%A
+
+[Source]
+Type=http-file
+Path=https://download.example.com/
+MatchPattern=foobarOS_$v_$u.root.xz
+
+[Target]
+Type=partition
+Path=auto
+MatchPattern=foobarOS_$v
+MatchPartitionType=root
+PartitionFlags=0
+PartitionReadOnly=1</programlisting></para>
+
+      <para>The above defines a matching transfer definition for the root file system.</para>
+
+      <para><programlisting># /usr/lib/sysupdate.d/70-kernel.conf
+[Transfer]
+ProtectVersion=%A
+
+[Source]
+Type=http-file
+Path=https://download.example.com/
+MatchPattern=foobarOS_$v.efi.xz
+
+[Target]
+Type=file
+Path=/efi/EFI/Linux
+MatchPattern=foobarOS_$v+$l-$d.efi \
+             foobarOS_$v+$l.efi \
+             foobarOS_$v.efi
+Mode=0444
+TriesLeft=3
+TriesDone=0
+InstancesMax=2</programlisting></para>
+
+        <para>The above installs a unified kernel image into the ESP (which is mounted to
+        <filename>/efi/</filename>), as per <ulink url="https://systemd.io/BOOT_LOADER_SPECIFICATION">Boot
+        Loader Specification</ulink> Type #2. This defines three possible patterns for the names of the
+        kernel images images, as per <ulink url="https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT">Automatic Boot
+        Assessment</ulink>, and ensures when installing new kernels, they are set up with 3 tries left. No
+        more than two parallel kernels are kept.</para>
+
+        <para>With this setup the web server would serve the following files, for a hypothetical version 7 of
+        the OS:</para>
+
+        <itemizedlist>
+          <listitem><para><filename>SHA256SUMS</filename> – The manifest file, containing available files and their SHA256 hashes</para></listitem>
+          <listitem><para><filename>SHA256SUMS.gpg</filename> – The detached cryptographic signature for the manifest file</para></listitem>
+          <listitem><para><filename>foobarOS_7_8b8186b1-2b4e-4eb6-ad39-8d4d18d2a8fb.verity.xz</filename> – The Verity image for version 7</para></listitem>
+          <listitem><para><filename>foobarOS_7_f4d1234f-3ebf-47c4-b31d-4052982f9a2f.root.xz</filename> – The root file system image for version 7</para></listitem>
+          <listitem><para><filename>foobarOS_7_efi.xz</filename> – The unified kernel image for version 7</para></listitem>
+        </itemizedlist>
+
+        <para>For each new OS release a new set of the latter three files would be added, each time with an
+        updated version. The <filename>SHA256SUMS</filename> manifest should then be updated accordingly,
+        listing all files for all versions that shall be offered for download.</para>
+    </example>
+
+    <example>
+      <title>Updates for Plain Directory Container Image</title>
+
+      <para><programlisting>
+[Source]
+Type=http-tar
+Path=https://download.example.com/
+MatchPattern=myContainer_$v.tar.gz
```

Comment:
Hmm, so the idea is that this scrapes the page and matches any links which match MatchPattern? Isn't that very fragile?

Reply:
"the page" is a SHA256SUMS file — not an HTML file. And that format is very well established already, many distros ship that and automatically validate it, because it's what GNU sha256sum tool generates and validates. i.e. `sha256sum -c` is the validator. That makes the format very well defined actually, there are plenty tools out there even outside of the GNU world, that deal with it.

So yes, the trick I am using here is that SHA256SUMS is not just good for validating check sums — it also doubles as a well-defined form of directory listing — an authenticated one on top.